src/main/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java

```diff
@@ -164,33 +175,39 @@ public OpticalDuplicateFinder(final String readNameRegex, final int opticalDupli
                 // and this loop seemed to kill the JVM for some reason. This logging statement (and the one in the
                 // loop below) solved the problem.
                 if (logProgress) progressLoggerForKeeper.record(String.format("%d", other.getReadGroup()), other.getX());
+                if (closeEnough(thisLoc, other, distance)) {
+                    opticalDistanceRelationGraph.addEdge(i, j);
+                }
             }
         }
-        if (logProgress) log.debug("Done with comparing to keeper, now the rest.");
+        if (logProgress) {
+            log.debug("Finished building adjacency graph for duplicate group, moving onto clustering");
+        }
 
-        // Now go through and do each pairwise comparison not involving the actualKeeper
-        for (int i = 0; i < length; ++i) {
-            final PhysicalLocation lhs = list.get(i);
-            if (lhs == actualKeeper) continue; // no comparisons to actualKeeper since those are all handled above
+        // Keep a map of the reads and their cluster assignments
+        final Map<Integer, Integer> opticalDuplicateClusterMap = opticalDistanceRelationGraph.cluster();
+        final Map<Integer, Integer> clusterToRepresentativeRead = new HashMap<>();
 
+        // Specially mark the keeper as specifically not a duplicate if it exists
+        if (keeperIndex >= 0) {
+            clusterToRepresentativeRead.put(opticalDuplicateClusterMap.get(keeperIndex),keeperIndex);
+        }
+
+        for (Map.Entry<Integer, Integer> entry : opticalDuplicateClusterMap.entrySet()) {
             // logging here for same reason as above
-            if (logProgress) progressLoggerForRest.record(String.format("%d", lhs.getReadGroup()), lhs.getX());
-
-            for (int j = i + 1; j < length; ++j) {
-                final PhysicalLocation rhs = list.get(j);
-                if (rhs == actualKeeper) continue; // no comparisons to actualKeeper since those are all handled above
-                if (opticalDuplicateFlags[i] && opticalDuplicateFlags[j]) continue; // both already marked, no need to check
-
-                if (closeEnough(lhs, rhs, distance)) {
-                    // At this point we want to mark either lhs or rhs as duplicate. Either could have been marked
-                    // as a duplicate of the keeper (but not both - that's checked above), so be careful about which
-                    // one to now mark as a duplicate.
-                    final int index = opticalDuplicateFlags[j] ? i : j;
-                    opticalDuplicateFlags[index] = true;
-                }
+            Integer recordIndex = entry.getKey();
```

Comment:
Use unboxed type `int` instead of `Integer` unless this can contain `null`.
```suggestion
            int recordIndex = entry.getKey();
```